### ISSUE TITLE
Huge performance issue with force-indexed usertypes.

### ISSUE THREAD

**Issue by OrfeasZ** [ThePhD/sol2#220]
We all know that sol2 is fast (**very** fast in some cases), however the currently written set of usertype benchmarks only assess a few edge cases where fast performance could be misinterpreted. What I mean by that is that even though for instance performance might be amazing when evaluating a single usertype variable/function access, there could be cases where performance greatly degrades when having multiple of such, accessed in random order.

I came across this observation while working on introducing member inheritance support for sol2. Even though benchmarks for any `simple_usertype` and function-only `usertype` registrations should provide similar results for a larger number of defined members, the iterative approach followed by force-indexed (`mustindex`) usertypes (ie. usertypes that also have member variables following the `obj.var` accessor format) could hinder performance, especially when attempting to access members present near the end of the usertype declaration. Below I've attached some benchmarks I've written in order to showcase this issue. I should note that I have not included `plain C` benchmarks as the implementation could vary greatly based on who's writing them.

Getting and setting 4 variables in non-sequential order:
![](http://i.nofate.me/UkKApWPQ2M.png)

Getting and setting 50 variables in non-sequential order:
![](http://i.nofate.me/p04A6tNjFW.png)

Getting and setting the last defined variable of a usertype with 50 defined variables:
![](http://i.nofate.me/g1Pml2IMYl.png)

As you can see, sol2 performance suffers a huge drop compared to every other library that supports this feature. I'll be making a fork of your lua-bench repo with my benchmarks attached so you can also evaluate them yourself. I'll update this issue when I do so.

**Comment by OrfeasZ** [comment]
Here are the same benchmarks but with `simple_usertype` instead:

Getting and setting 4 variables in non-sequential order:
![](http://i.nofate.me/0ov45mxV6H.png)

Getting and setting 50 variables in non-sequential order:
![](http://i.nofate.me/oPp5Q4guKZ.png)

Getting and setting the last defined variable of a usertype with 50 defined variables:
![](http://i.nofate.me/yKk2xLE0h9.png)

It should be noted that all benchmarks were performed on a i7-5930K @ 4.20GHz

**Comment by OrfeasZ** [comment]
You can find all my benchmarks [here](https://github.com/OrfeasZ/lua-bench/tree/complex-bench).

**Comment by ThePhD** [comment]
I knew linear search would come back to bite me in the ass.

Doing compile-time sorting / mapping is just a tad hard, though, so I'm going to have to figure out some other clever way of working with things. Either that, or I fold all implementations to use `simple_usertype` silently...

**Comment by ThePhD** [comment]
Performance data to help inform my decisions for the next good while. I benchmarked the cost of string lookup under various paradigms.

TL;DR: `unordered_map` is kind for any kind of non-trivial data:

Using the nonius beta:

*50 strings*
![Hot](http://i.imgur.com/Csdr540.png)

*10 strings*
![Damn](http://i.imgur.com/ToXeRdU.png)

I guess I'll need to figure out how to `unordered_map` and still keep stuff at compile time. Probably `unordered_map<string, std::size_t>` or something.

**Comment by ThePhD** [comment]
Still generating numbers but I think this should be more or less taken care of with https://github.com/ThePhD/sol2/releases/tag/v2.14.2 -- I just have to make sure of it...

**Comment by ThePhD** [comment]
The numbers are all better now.

![Clean](https://github.com/ThePhD/lua-bench/blob/master/lua%20-%20results/lua%20bench%20graph%20-%20many%20member%20variables.png?raw=true)

![Nice](https://github.com/ThePhD/lua-bench/blob/master/lua%20-%20results/lua%20bench%20graph%20-%20many%20member%20variables.png?raw=true)
